File: FONTS/ClouseauApp/APP/APP/App_Front/models/Game.js


let Game = (function () {
    let privateProps = new WeakMap();

    class Game {
        constructor(data) {
            let random = (function (min, max) {
                return function (min, max) {
                    min = Math.ceil(min);
                    max = Math.floor(max)+.99;
                    return Math.floor(Math.random() * (max - min)) + min;
                }
            })();

            privateProps.set(this, {
                solution: new Teory(random(1, data.suspect.length)
                                   , random(1, data.place.length)
                                   , random(1, data.weapon.length)
                    ),
                random: random
            });
        }

        submit(teory) {
            let ret = privateProps.get(this).solution.valid(teory.suspect, teory.place, teory.weapon);
            if (ret.length > 0) {
                let test = ret[privateProps.get(this).random(1, ret.length)];
                if (!!test) {
                    return test;
                } else {
                    return 1;
                }
            }
            else {
                return 0;
            }
        }


    }

    return Game;
})();

File: SCRIPTS/ESBOÇO_MOTOR/console-motor-jogo-eniwine-01/console-motor-jogo-eniwine-01/Models/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Models
{

    public class Game
    {
        private Rules rules;
        private Teory solution;

        public Game()
        {
            rules = new Rules();
        }
        public Game(int players)
        {
            rules = new Rules();
        }


        public void Start()
        {
            solution = new Teory()
            {
                Suspeito = lottery(rules.Suspeitos),
                Local = lottery(rules.Locais),
                Arma = lottery(rules.Armas)
            };

        }

        private int lottery(int begin, int end)
        {
            var rnd = new Random();
            return rnd.Next(begin, end);
        }


        private KeyValuePair<int, string> lottery(Dictionary<int, string> collection)
        {
            var begin = collection.Min(e => e.Key);
            var end = collection.Max(e => e.Key);
            return collection.FirstOrDefault(e=>e.Key== lottery(begin, end));
        }

        public int Submit(Teory teory)
        {
            var ret =new List<int> { };
            if (solution.Suspeito.Key != teory.Suspeito.Key)
            {
                ret.Add(1);
            }

            if (solution.Local.Key != teory.Local.Key)
            {
                ret.Add(2);
            }

            if (solution.Arma.Key != teory.Arma.Key)
            {
                ret.Add(3);
            }

            if (ret.Count>0)
            {
                return ret[lottery(1, ret.Count) -1];
            }
            else {
                return 0;
            }

        }






    }
}